src/sqlfluff/core/rules/functional/segments.py

```diff
@@ -0,0 +1,140 @@
+            for child in s.segments:
+                if cp(child):
+                    child_segments.append(child)
+        return Segments(self.templated_file, *child_segments)
+
+    def first(self, *predicates) -> Optional["Segments"]:
+        """Returns the first segment (if any) that satisfies the predicates."""
+        cp = _CompositePredicate(*predicates)
+        for s in self:
+            if cp(s):
+                return Segments(self.templated_file, s)
+        # If no segment satisfies "predicates", return "None".
+        return None
+
+    def last(self, *predicates) -> Optional["Segments"]:
+        """Returns the last segment (if any) that satisfies the predicates."""
+        cp = _CompositePredicate(*predicates)
+        for s in reversed(self):
+            if cp(s):
+                return Segments(self.templated_file, s)
+        # If no segment satisfies "predicates", return "None".
+        return None
+
+    def apply(self, fn: Callable[[BaseSegment], Any]) -> List[Any]:
+        """Apply function to every item."""
+        return [fn(s) for s in self]
+
+    def select(
+        self,
+        select_if: Optional[Sequence[Predicate]] = None,
+        loop_while: Optional[Sequence[Predicate]] = None,
+        start_seg: Optional[BaseSegment] = None,
+        stop_seg: Optional[BaseSegment] = None,
+    ) -> "Segments":
+        """Retrieve range/subset.
+
+        NOTE: Iterates the segments BETWEEN start_seg and stop_seg, i.e. those
+        segments are not included in the loop.
+        """
+        start_index = self.index(start_seg) if start_seg else -1
+        stop_index = self.index(stop_seg) if stop_seg else len(self)
+        buff = []
+        cp_select_if = None
+        if select_if:
+            cp_select_if = _CompositePredicate(*select_if)
+        cp_loop_while = None
+        if loop_while:
+            cp_loop_while = _CompositePredicate(*loop_while)
+        for seg in self[start_index + 1 : stop_index]:
+            if cp_loop_while and not cp_loop_while(seg):
+                break
+            if not cp_select_if or cp_select_if(seg):
+                buff.append(seg)
+        return Segments(self.templated_file, *buff)
+
+
+class _CompositePredicate:
+    def __init__(self, *predicates: Predicate):
+        self.type_names: List[str] = []
+        self.other = []
+        for p in predicates:
+            if isinstance(p, str):
+                self.type_names.append(p)
+            else:
+                self.other.append(p)
+
+    def __call__(self, segment: BaseSegment) -> bool:
+        if self.type_names and not segment.is_type(*self.type_names):
+            return False
+
+        for p in self.other:
+            if isinstance(p, type):
+                # Check segment class
+                if not isinstance(segment, p):  # pragma: no cover
+                    return False
+            elif not p(segment):  # Arbitrary function
+                return False
+        return True
```

Comment:
@barrywhart just a general thought from playing around further with the predicates:

We've added in the `not_()`, `and_()`, `or_()` predicates to allow us to combine multiple function predicates. However, predicates also allows types to be supplied as strings. The issue this raises for me is that we still can't combine a str type predicate and a function predicate using these boolean predicates.
e.g. can't do something like `segpred.and_("object_reference", segpred.is_expandable)`.

Think we should refine our definition of predicates to solve this and simplify things a bit:
1. make predicates only take in `Callable[[BaseSegment], bool]`.
2. replace the types logic with a common predicate:
```python
def is_type(
    *seg_type: str
) -> Callable[[BaseSegment], bool]:  # pragma: no cover
    """Returns a function that determines if the segment is of the supplied types."""
    def _(segment: BaseSegment):
        return segment.is_type(*seg_type)
    return _
```
(use a closure to generate the predicate)
3. the types can easily be combined with other predicates and keeps the API consistent: `segpred.and_(segpred.is_type("object_reference"), segpred.is_expandable)`

Let me know what you think, but I quite like the consistency 👍

Reply:
We can also drop the composite predicates classes in favour of this as well

Reply:
Sounds good to me. In my early commits, I didn't think through the usage really carefully, was just "sketching" something to try out.